efaCloud

nicmichael opened the pull request:
efaCloud von Martin am 17.05.2020

nicmichael commented on de/nmichael/efa/gui/util/EfaMenuButton.java:
```diff
@@ -174,6 +175,12 @@ public boolean isSeparator() {
                     International.getStringWithMnemonic("Backups"),
                     BaseFrame.getIcon("menu_backup.png")));
         }
+        if (admin == null || admin.isAllowedCreateBackup()) {
```
Instead of backup permission, should be maybe check for isAllowedAdministerProjectLogbook()?

nicmichael commented on de/nmichael/efa/gui/util/EfaMenuButton.java:
```diff
@@ -464,6 +471,17 @@ private static boolean menuAction(BaseFrame parentFrame, BaseDialog parentDialog
 
+        if (action.equals(BUTTON_EFACLOUD)) {
+            if (admin == null ||
+                    (!admin.isAllowedCreateBackup() && !admin.isAllowedRestoreBackup())) {
```
See above: isAllowedAdministerProjectLogbook() instead?

nicmichael commented on de/nmichael/efa/data/efacloud/TableBuilder.java:
```diff
@@ -0,0 +1,520 @@
+    public HashMap<String, RecordTypeDefinition> recordTypes = new HashMap<>();
+    // when starting the application, data will be synchronized by loading them from the server
+    // to show the load progress, the GUI context is needed. It can be put here
+    public EfaBaseFrame guiBaseFrameOnAppLoading = null;
```
This is currently only being set from EfaBaseFrame (under some conditions which aren't clear to me), but not from EfaBoathouseFrame. Does it need to be set from both?

tfyh replied:
EfaBoathouseFrame does write any structure information into the efaCloud DB, but only records. It therefore does not need that HashMap. Nonetheless I will have a look. Inconsistencies shall be avoided, I agrre.

tfyh replied:
guiBaseFrameOnAppLoading is used by the de.nmichael.efa.data.efacloud.EfaCloudSynch synchronization process as parentDialog in the synchronization progress display by de.nmichael.efa.gui.ProgressDialog. Now the synchronization can only be triggered from the BaseFrame menu, as it was not implemented by me in the efaBoatHouse menu. I just didn't deal with it for now. If it shall be usable also in that context via admin login, this needs still to be done.

nicmichael commented on de/nmichael/efa/data/ProjectRecord.java:
```diff
@@ -786,15 +797,16 @@ public int getBoathouseId() {
                                     + International.getString("remote") + ")"
                                     : International.getString("Speicherort"))));
-                    ((ItemTypeString) item).setEnabled(getStorageType() != IDataAccess.TYPE_FILE_XML);
-                    ((ItemTypeString) item).setNotNull(getStorageType() == IDataAccess.TYPE_DB_SQL);
```
Hmm, looks like this line was incorrect to begin with.

tfyh commented on de/nmichael/efa/data/efacloud/TableBuilder.java:
```diff
@@ -0,0 +1,520 @@
+     * @param efa2RecordType type of data record to be removed, e.g. efa2persons
+     */
+    public void removeDataRecord(String efa2RecordType) {
+        String recordType = efaCloudTableNames.get(efa2RecordType);
```
Sure there is an issue with completeness, if new efa record types are created. The entire duplication of the structure is sub-optimal. Nevertheless I'd leave it for cleansing once we got it running.

tfyh commented on de/nmichael/efa/data/efacloud/TableBuilder.java:
```diff
@@ -0,0 +1,520 @@
+     * There are essentially two namespaces for efa2 records the efa2 record type name like
+     * efa2persons and the record's table class name, like Persons. This here maps the name spaces
+     */
+    public static final HashMap<String, String> efaCloudTableNames = new HashMap<>();
```
I'll have a look into it. When I started exercise I was not clear about the different ways a record is referenced and created Hashmaps to be able to refer whatever was needed. Also an exercise for cleansing the design, You're right.

tfyh replied:
I finally looked into it. The reason is simple: we do have a local XML data base where records have a StorageObject to which they are stored, e.g. de.nmichael.efa.core.config.EfaTypes. The record definition is in de.nmichael.efa.core.config.EfaTypeRecord and to reference the StorageObject there is are getter & setter in IDataAccess (set/getStorageType()), e.g. 'efa2types'. Now the SQL DB of the efaCloud server uses tables to store the record. Their names are needed for the client-server communication. Now the question is, why I did not use the "storageType" namespace. Well, I just didn't understand what best to use when building the server tables. I will change it now to the StorageType.

nicmichael replied:
I tried to remove the mapping and just used the plain storage type names as table names, but this resulted in some ArrayOutOfBoundsExceptions in the synchronization. They seemed to be coming from the server... but I haven't debugged it more. I'm leaving it as is for now. But it's maybe something that should be changed before this feature becomes public as it would affect the table names on the server.

tfyh replied:
I am going to do that change today (it is the morning of 4th Oct here already). It does affect the server so it must be done on both sides to work. Please go ahead and merge your code changes. I will then join the git type of collaboration as proposed.

nicmichael commented:
@tfyh I've pushed another commit with a few smaller changes. Mostly a few cosmetic ones.

Among them:

I've added a new option "Experimentelle Funktionalitäten aktivieren" (default off). You'll have to go to the efa Configuration dialog, activate "Expert mode" and then (on the default tab) enable it. Only then you'll see the EfaCloud options.

I also modified some of the internationalized strings. The convention for the text inside them is that it must be German (no English), so \n inside the strings, and ideally short individual sentences with as much reuse of text blocks as possible. Every of these strings potentially has to be translated into other languages (and efa currently supports quite a few), so I'm trying to keep the number of strings as low as can be (currently more than 1700). There's some very technical (internal) error messages you wrote in English. I suppose those aren't meant for the ordinary user and don't need translation, so I replaced the International.getMessage with String.format for them. 

I've also added your name to the list of contributors.

Take a look at the current code. Maybe you can give it another try whether nothing broke? If it's ok, I'll integrate it into master.

nicmichael commented on de/nmichael/efa/gui/EfaCloudSynchDialog.java:
```diff
@@ -0,0 +1,206 @@
+                    return;
+                }
+                // Start the server upload.
+                EfaCloudSynch.runEfaCloudSynchTask(this, EfaCloudSynch.Mode.upload, true, false, 0L,
```
@tfyh This may need some improvement. In my testing, if often happened that the synchronization failed (for whatever reason, like me entering the wrong password or URL). While that will print an error, it doesn't undo the previous steps to update the project. I guess you have to first update the project before kicking off the synchronization (otherwise I'd just switch the order). So I guess we'd need to explicitly reset the storage type to what it used to be in case of fatal synchronization errors.

tfyh replied:
Yes. I will simplify the synchronisation. It is too complex right now for a robust operation.

tfyh commented:
Thanks for the review and changes. I blocked this Sunday for a full revision of the efacloud package.

nicmichael commented:
@tfyh I just merged the changes into master: https://github.com/nicmichael/efa/commit/c3b9a9ad1c5f559c1448b8bcf76c5dedfdfdfcd5

For future code changes, please directly send me pull requests from your fork.
